### PR TITLE
Raise error when read and write happens to the same directory location in the same task graph

### LINKED ISSUE #7466: Read/write to same file using  `.to_parquet(overwrite=True)` in the same task graph should raise an error
Performing the operation `.to_parquet` produces an error if the Dask dataframe is loaded from a Dask file vs. if it is created within the code and not read from file.

**What happened**: **Way 1** of the code below works as expected - it correctly saves the output to parquet. **Way 2** of the below code, does NOT work as expected and fails with `FileNotFoundError: [Errno 2] No such file or directory: 'error.parquet/part.1.parquet'`

**What you expected to happen**: I expected **Way 2** to work as **Way 1**

**Minimal Complete Verifiable Example**:

```python
import numpy as np
import pandas as pd
import dask.dataframe as dd

# -------------
# way 1 - works
# -------------
print('way 1 - start')
A = np.random.rand(200,300)
cols = np.arange(0, A.shape[1])
cols = [str(col) for col in cols]
df = pd.DataFrame(A, columns=cols)
ddf = dd.from_pandas(df, npartitions=11)

# compute and resave
ddf.drop(cols[0:11], axis=1)
dd.to_parquet(
	ddf, 'error.parquet', engine='auto', compression='default',
	write_index=True, overwrite=True, append=False)
print('way 1 - end')

# ----------------------
# way 2 - does NOT work 
# ----------------------
print('way 2 - start')
ddf = dd.read_parquet('error.parquet')

# compute and resave
ddf.drop(cols[0:11], axis=1)
dd.to_parquet(
	ddf, 'error.parquet', engine='auto', compression='default',
	write_index=True, overwrite=True, append=False)
print('way 2 - end')
```

**Environment**:

- Dask version: dask==2021.3.0
- Python version: 3.8
- Operating System: Mac OS Big Sur 11.2.1
- Install method (conda, pip, source): pip

### PR DESCRIPTION
**Pull request by Madhu94**
- [x] Closes #7466
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

I had a few queries, marking as draft till then.
* The format of `parts`. The fastparquet engine seems to accept [two formats](https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/dataframe/io/parquet/fastparquet.py#L750) - one where the path is the first element of the `parts` tuple and another where the path is part of `kwargs`. I can't seem to find usage of the second format anywhere. Could you point me to an example so I can handle that case too?

### CODE REVIEW

**Comment by jsignell** [comment]
Pinging @martindurant for fastparquet expertise :)

**Comment by martindurant** [comment]
This certainly implements @crusaderky 's suggestion to solve the problem and gives tests showing that it works. I am *slightly* uneasy about inspecting the graph like this, I fell like it's likely to break in the future. In the current implementation, I think it might fail when:
- the root parquet directory contains a nested set of directories (as is typical for path-based partitioning)
- the read is from a single file rather than a directory, but then attempts to write back to that file. I imagine you end up with an empty directory with the same name and no data.
If those points are also tested, I'd be OK to merge this despite my unease.

As an aside, we might also want to explicitly think about the case of overwrite=False, where the writes might overlap some or all existing files, perhaps resulting in incomplete writes.

To the comment about writing and moving only on completion: this is tricky on object stores. Instead, the backends usually support confirm/abort for creating the individual files, which would be tricky to fit into the framework we have now. There is discussion somewhere on s3fs about the possibility of transactional writing across a distributed cluster.

Note that this is not really a fastparquet thing.

**Comment by AlanGanem** [comment]
Hey! I'm facing an issue that might be solved by this PR. Do you guys have any idea when this will be merged or released?

**Comment by martindurant** [comment]
There were some unresolved conversation points, above. It would require more work from @Madhu94 (or someone else) to resolve these, which may not be trivial given the complexity of the situation.

**Comment by ncclementi** [comment]
@Madhu94 Checking in here, are you still interested in working on this? There were a couple of concerns raised in https://github.com/dask/dask/pull/7519#issuecomment-813459590 

Would you be able to take a look at those and work on them?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by bryanwweber** [comment]
@martindurant This looks like it was updated by @Madhu94 and needs another review. Are you able to help here? cc @rjzamora Thank you!

**Comment by martindurant** [comment]
(I updated the formatting)

This seems like a reasonable thing to do. @rjzamora probably could do it more concisely without iterating over all the tasks of the first layer; but I think this is OK. An additional improvement may be to skip the check if the directory is empty.

File: dask/dataframe/io/parquet/core.py [hunk 1]
```diff
@@ -579,6 +581,18 @@ def to_parquet(
...
             raise ValueError("Cannot use both `overwrite=True` and `append=True`!")
         if fs.exists(path) and fs.isdir(path):
+            # Ref: https://github.com/dask/dask/issues/7466
+            # Check that we are not writing to the same location we are reading from
+            graph = df.__dask_graph__()
+            for layer_name, layer in graph.layers.items():
+                if layer_name.startswith("read-parquet-"):
+                    for k, v in layer.items():
+                        # check the arguments of the callable to extract the path
+                        if os.path.dirname(v[1]["piece"][0]) == path:
```
Comment (martindurant) [inline dask/dataframe/io/parquet/core.py:590]: Actually, this should probably not be `os.path`, because all paths ought to be posix-like.

**Comment by jcrist** [comment]
This has since been fixed in https://github.com/dask/dask/pull/9124, closing.

**Pull request closed without merging**